apps/sombrero/reframe_sombrero.py

```diff
@@ -7,25 +7,102 @@
 import os.path as path
 import reframe as rfm
 import reframe.utility.sanity as sn
+from modules.reframe_extras import scaling_config
+import itertools as it
+from enum import IntEnum
+
+
+class SombreroCase:
+    class Idx(IntEnum):
+        # See _flatten_nested_case() and generate()
+        strong_or_weak = 0
+        size = 1
+        partition = 2
+        nprocesses = 3
+        nprocesses_per_node = 4
+        theory_id = 5
+
+    @staticmethod
+    def _flatten_nested_case(nested_case):
+        # See Idx and generate()
+        return (
+            nested_case[0],  # strong/weak
+            nested_case[1],  # size
+            *nested_case[2],  # input from scaling config
+            nested_case[3])  # theory_id
+
+    @staticmethod
+    def _check_nprocesses(np):
+        '''
+        Make sure the number of processes is 2^n or 2^n x 3
+        '''
+        assert np > 0
+        if np % 2 == 1:
+            return np in [1, 3]
+        else:
+            return SombreroCase._check_nprocesses(np // 2)
+
+    @staticmethod
+    def _case_filter(case):
+        strong_or_weak = case[SombreroCase.Idx.strong_or_weak]
+        size = case[SombreroCase.Idx.size]
+        nprocesses = case[SombreroCase.Idx.nprocesses]
+
+        if not SombreroCase._check_nprocesses(nprocesses):
+            return False
+        if strong_or_weak == "weak" and size != "medium":
+            return False
+
+        return True
+
+    @staticmethod
+    def generate(scaling_config):
+        theory_ids = range(1, 7)
+        sizes = ['small', 'medium', 'large', 'very_large']
+        strong_or_weak = ['strong', 'weak']
+
+        # See Idx and _flatten_nested_case()
+        nested_cases = it.product(strong_or_weak, sizes, scaling_config(),
+                                  theory_ids)
+
+        cases = map(SombreroCase._flatten_nested_case, nested_cases)
+
+        filtered = filter(SombreroCase._case_filter, cases)
+
+        filtered_list = list(filtered)
+        for case in filtered_list:
+            assert len(case) == 6
+        return filtered_list
 
 
 @rfm.simple_test
 class SombreroBenchmark(rfm.RegressionTest):
-    valid_systems = ['*']
-    valid_prog_environs = ['*']
-    build_system = 'Spack'
-    executable = 'sombrero.sh'
-    executable_opts = ['-w', '-s', 'medium', '-n', '4']
-    time_limit = '10m'
-    reference = {
-        '*': {
-            'flops': (30, -0.5, 0.5, 'Gflops/seconds'),
+    params = parameter(SombreroCase.generate(scaling_config))
```

Comment:
I'm a bit lost: where does `parameter` come from?

Reply:
`parameter` is a built in type of reframe:
https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html?highlight=parameter#built-in-types